diff --git java/android/support/v4/app/FragmentActivity.java java/android/support/v4/app/FragmentActivity.java
index ad57bb8..896493b 100644
--- java/android/support/v4/app/FragmentActivity.java
+++ java/android/support/v4/app/FragmentActivity.java
@@ -164,7 +164,7 @@ public class FragmentActivity extends Activity {
                 Log.w(TAG, "Activity result no fragment exists for index: 0x"
                         + Integer.toHexString(requestCode));
             } else {
-                frag.onActivityResult(requestCode&0xffff, resultCode, data);
+                rOnActivityResult(frag, requestCode & 0xffff, resultCode, data);
             }
             return;
         }
@@ -172,6 +172,17 @@ public class FragmentActivity extends Activity {
         super.onActivityResult(requestCode, resultCode, data);
     }
 
+    private void rOnActivityResult(Fragment fragment, int requestCode, int resultCode, Intent data) {
+        fragment.onActivityResult(requestCode, resultCode, data);
+        if(fragment.mChildFragmentManager != null && fragment.mChildFragmentManager.mActive != null) {
+            for(Fragment childFragment : fragment.mChildFragmentManager.mActive) {
+                if(childFragment != null) {
+                    rOnActivityResult(childFragment, requestCode, resultCode, data);
+                }
+            }
+        }
+    }
+
     /**
      * Take care of popping the fragment back stack or finishing the activity
      * as appropriate.
diff --git java/android/support/v4/app/FragmentManager.java java/android/support/v4/app/FragmentManager.java
index 60a6367..5319c69 100644
--- java/android/support/v4/app/FragmentManager.java
+++ java/android/support/v4/app/FragmentManager.java
@@ -873,6 +873,7 @@ final class FragmentManagerImpl extends FragmentManager {
                 case Fragment.INITIALIZING:
                     if (DEBUG) Log.v(TAG, "moveto CREATED: " + f);
                     if (f.mSavedFragmentState != null) {
+                        f.mSavedFragmentState.setClassLoader(mActivity.getClassLoader());
                         f.mSavedViewState = f.mSavedFragmentState.getSparseParcelableArray(
                                 FragmentManagerImpl.VIEW_STATE_TAG);
                         f.mTarget = getFragment(f.mSavedFragmentState,
@@ -916,7 +917,7 @@ final class FragmentManagerImpl extends FragmentManager {
                                 f.mSavedFragmentState), null, f.mSavedFragmentState);
                         if (f.mView != null) {
                             f.mInnerView = f.mView;
-                            f.mView = NoSaveStateFrameLayout.wrap(f.mView);
+                            f.mView = NoSaveStateFrameLayout.wrap(f, f.mView);
                             if (f.mHidden) f.mView.setVisibility(View.GONE);
                             f.onViewCreated(f.mView, f.mSavedFragmentState);
                         } else {
@@ -943,7 +944,7 @@ final class FragmentManagerImpl extends FragmentManager {
                                     f.mSavedFragmentState), container, f.mSavedFragmentState);
                             if (f.mView != null) {
                                 f.mInnerView = f.mView;
-                                f.mView = NoSaveStateFrameLayout.wrap(f.mView);
+                                f.mView = NoSaveStateFrameLayout.wrap(f, f.mView);
                                 if (container != null) {
                                     Animation anim = loadAnimation(f, transit, true,
                                             transitionStyle);
diff --git java/android/support/v4/app/NoSaveStateFrameLayout.java java/android/support/v4/app/NoSaveStateFrameLayout.java
index d0ebb96..c9f9480 100644
--- java/android/support/v4/app/NoSaveStateFrameLayout.java
+++ java/android/support/v4/app/NoSaveStateFrameLayout.java
@@ -28,7 +28,20 @@ import android.widget.FrameLayout;
  * so instead we insert this between the view and its parent.
  */
 class NoSaveStateFrameLayout extends FrameLayout {
+    static interface NoSaveStateFrameLayoutCreator {
+        public ViewGroup create(Fragment fragment, View child);
+    }
+	
+    static NoSaveStateFrameLayoutCreator sNoSaveStateFrameLayoutCreator;
+
     static ViewGroup wrap(View child) {
+        return wrap(null, child);
+    }
+
+    static ViewGroup wrap(Fragment fragment, View child) {
+        if (sNoSaveStateFrameLayoutCreator != null) {
+            return sNoSaveStateFrameLayoutCreator.create(fragment, child);
+        }
         NoSaveStateFrameLayout wrapper = new NoSaveStateFrameLayout(child.getContext());
         ViewGroup.LayoutParams childParams = child.getLayoutParams();
         if (childParams != null) {
diff --git java/android/support/v4/internal/view/SupportContextMenu.java java/android/support/v4/internal/view/SupportContextMenu.java
new file mode 100644
index 0000000..86bd22b
--- /dev/null
+++ java/android/support/v4/internal/view/SupportContextMenu.java
@@ -0,0 +1,73 @@
+package android.support.v4.internal.view;
+
+import android.graphics.drawable.Drawable;
+import android.view.ContextMenu;
+import android.view.View;
+
+/**
+ * Extension of {@link android.view.Menu} for context menus providing functionality to modify
+ * the header of the context menu.
+ * <p>
+ * Context menus do not support item shortcuts and item icons.
+ * <p>
+ * To show a context menu on long click, most clients will want to call
+ * {@link android.app.Activity#registerForContextMenu} and override
+ * {@link android.app.Activity#onCreateContextMenu}.
+ *
+ * <div class="special reference">
+ * <h3>Developer Guides</h3>
+ * <p>For information about creating menus, read the
+ * <a href="{@docRoot}guide/topics/ui/menus.html">Menus</a> developer guide.</p>
+ * </div>
+ */
+public interface SupportContextMenu extends SupportMenu, ContextMenu {
+    /**
+     * Sets the context menu header's title to the title given in <var>titleRes</var>
+     * resource identifier.
+     *
+     * @param titleRes The string resource identifier used for the title.
+     * @return This ContextMenu so additional setters can be called.
+     */
+    public SupportContextMenu setHeaderTitle(int titleRes);
+
+    /**
+     * Sets the context menu header's title to the title given in <var>title</var>.
+     *
+     * @param title The character sequence used for the title.
+     * @return This ContextMenu so additional setters can be called.
+     */
+    public SupportContextMenu setHeaderTitle(CharSequence title);
+
+    /**
+     * Sets the context menu header's icon to the icon given in <var>iconRes</var>
+     * resource id.
+     *
+     * @param iconRes The resource identifier used for the icon.
+     * @return This ContextMenu so additional setters can be called.
+     */
+    public SupportContextMenu setHeaderIcon(int iconRes);
+
+    /**
+     * Sets the context menu header's icon to the icon given in <var>icon</var>
+     * {@link Drawable}.
+     *
+     * @param icon The {@link Drawable} used for the icon.
+     * @return This ContextMenu so additional setters can be called.
+     */
+    public SupportContextMenu setHeaderIcon(Drawable icon);
+
+    /**
+     * Sets the header of the context menu to the {@link View} given in
+     * <var>view</var>. This replaces the header title and icon (and those
+     * replace this).
+     *
+     * @param view The {@link View} used for the header.
+     * @return This ContextMenu so additional setters can be called.
+     */
+    public SupportContextMenu setHeaderView(View view);
+
+    /**
+     * Clears the header of the context menu.
+     */
+    public void clearHeader();
+}
diff --git java/android/support/v4/view/ViewCompat.java java/android/support/v4/view/ViewCompat.java
index 84455a0..ef84f72 100644
--- java/android/support/v4/view/ViewCompat.java
+++ java/android/support/v4/view/ViewCompat.java
@@ -286,6 +286,7 @@ public class ViewCompat {
         public int getMeasuredState(View view);
         public int getAccessibilityLiveRegion(View view);
         public void setAccessibilityLiveRegion(View view, int mode);
+        public int[] mergeDrawableStates(int[] baseState, int[] additionalState);
         public int getPaddingStart(View view);
         public int getPaddingEnd(View view);
         public void setPaddingRelative(View view, int start, int top, int end, int bottom);
@@ -428,7 +429,25 @@ public class ViewCompat {
         }
 
         public int resolveSizeAndState(int size, int measureSpec, int childMeasuredState) {
-            return View.resolveSize(size, measureSpec);
+            int result = size;
+            int specMode = View.MeasureSpec.getMode(measureSpec);
+            int specSize = View.MeasureSpec.getSize(measureSpec);
+            switch (specMode) {
+                case View.MeasureSpec.UNSPECIFIED:
+                    result = size;
+                    break;
+                case View.MeasureSpec.AT_MOST:
+                    if (specSize < size) {
+                        result = specSize | MEASURED_STATE_TOO_SMALL;
+                    } else {
+                        result = size;
+                    }
+                    break;
+                case View.MeasureSpec.EXACTLY:
+                    result = specSize;
+                    break;
+            }
+            return result | childMeasuredState & MEASURED_STATE_MASK;
         }
 
         @Override
@@ -646,6 +665,17 @@ public class ViewCompat {
         public float getPivotY(View view) {
             return 0;
         }
+
+        @Override
+        public int[] mergeDrawableStates(int[] baseState, int[] additionalState) {
+            final int N = baseState.length;
+            int i = N - 1;
+            while (i >= 0 && baseState[i] == 0) {
+                i--;
+            }
+            System.arraycopy(additionalState, 0, baseState, i + 1, additionalState.length);
+            return baseState;
+        }
     }
 
     static class EclairMr1ViewCompatImpl extends BaseViewCompatImpl {
@@ -1633,6 +1663,10 @@ public class ViewCompat {
         IMPL.setAccessibilityLiveRegion(view, mode);
     }
     
+    public static int[] mergeDrawableStates(int[] baseState, int[] additionalState) {
+        return IMPL.mergeDrawableStates(baseState, additionalState);
+    }
+
     /**
      * Returns the start padding of the specified view depending on its resolved layout direction.
      * If there are inset and enabled scrollbars, this value may include the space
diff --git kitkat/android/support/v4/print/PrintHelperKitkat.java kitkat/android/support/v4/print/PrintHelperKitkat.java
index 89cc9e7..77d7ce6 100644
--- kitkat/android/support/v4/print/PrintHelperKitkat.java
+++ kitkat/android/support/v4/print/PrintHelperKitkat.java
@@ -45,7 +45,7 @@ import java.io.InputStream;
  */
 class PrintHelperKitkat {
     private static final String LOG_TAG = "PrintHelperKitkat";
-    // will be <= 300 dpi on A4 (8.3×11.7) paper (worst case of 150 dpi)
+    // will be <= 300 dpi on A4 (8.3x11.7) paper (worst case of 150 dpi)
     private final static int MAX_PRINT_SIZE = 3500;
     final Context mContext;
     BitmapFactory.Options mDecodeOptions = null;
